### PR TITLE
Make Webpacker the default JavaScript compiler for Rails 6

### PR DESCRIPTION
**Pull request by dhh**
Make Webpacker the default Javascript compiler for Rails 6 with the following changes:

- Webpacker gem is installed by default and webpacker:install is run by the Rails application generator.
- Action Cable channel generators will create ES6 stubs rather than use CoffeeScript
- Active Storage, Action Cable, Turbolinks, and Rails-UJS is loaded by a new application.js pack in app/javascript by default (unless any of the frameworks have been opted out).
- Active Storage, Action Cable, Turbolinks, and Rails-UJS npm modules are automatically listed as dependencies in the default package.json.
- All the JavaScript-related auxiliaries for Sprockets, like compression and uglifying, is no longer configured or included by default.
- No JavaScript stubs will be created by default when using the scaffold generators any more.

### CODE REVIEW

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -327,59 +324,26 @@ def convert_database_option_for_jruby
...
-      def javascript_runtime_gemfile_entry
-        comment = "See https://github.com/rails/execjs#readme for more supported runtimes"
-        if defined?(JRUBY_VERSION)
-          GemfileEntry.version "therubyrhino", nil, comment
-        elsif RUBY_PLATFORM =~ /mingw|mswin/
-          GemfileEntry.version "duktape", nil, comment
-        else
-          GemfileEntry.new "mini_racer", nil, comment, { platforms: :ruby }, true
+          [ GemfileEntry.version("turbolinks", "~> 5",
+             "Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks") ]
```
Comment (yhirano55) [inline railties/lib/rails/generators/app_base.rb]: Turbolinks is added on package.json, so it looks unnecessary to add turbolinks to Gemfile.

**Comment by dhh** [comment]
There’s a Ruby and a JS component. Both are needed.

File: railties/lib/rails/generators/app_base.rb (hunk 1 above)
Comment (yhirano55) [reply railties/lib/rails/generators/app_base.rb]: I see. Thanks. https://github.com/turbolinks/turbolinks-rails

**Comment by rainerborene** [comment]
@dhh any reason to keep the `javascript` folder name in singular? why not following the pattern of plural name like controllers, models and views?

**Comment by dhh** [comment]
Because JavaScript is a language, not a classification of a domain type,
like controllers or views or models. JavaScript has all those things.
Saying "JavaScripts" imply that it's a collection of isolated, independent
files, when it's not. It's the part of the application written in
JavaScript.

**Comment by dhh** [comment]
No, Sprockets is still to be used for SCSS (and other CSS processors) as
well as copying all other static assets.

**Comment by dwightwatson** [comment]
Out of curiousity, what is the argument to continue using Sprockets for CSS/static assets when Webpacker supports them by default out of the box?

**Comment by adrianthedev** [comment]
My `javascripts` folder already has the `images`, `js`, `packs`, `stylesheets` folders inside. Would be nice to use the old `assets` folder instead of the current `javascripts` to keep all assets.

**Comment by rainerborene** [comment]
> No, Sprockets is still to be used for SCSS (and other CSS processors) as well as copying all other static assets.

@dhh I see your point. But it's inevitable that these JavaScript packages will make use of their own internal styling (like React or something). So, you'll end up with styles in two separate places and from different "compilers". That could become a mess, don't you think?

**Comment by dhh** [comment]
I don’t think it’s a problem. You can have Webpack do your CSS as well, if you like. But the default path is not having to muck with that.

**Comment by frodsan** [comment]
@adrianthedev I think you can change the default path in `.webpacker.yml`:

```diff
default: &default
-  source_path: app/javascript
+ source_path: app/assets
```

**Comment by adrianthedev** [comment]
For sure I can do that @frodsan. But by doing it that it would change the default "rails" way.

This is the turning point when we can have that default path in `assets` and all generators and other packages point to that. 
Again, making it the rails way (and the consistent way) of keeping everything (js, images, css, etc.) in the `assets` folder and **giving the developer the choice of using Webpack or sprockets** for his CSS.
It seems as the perfect setup.

**Comment by coorasse** [comment]
Keep supporting both sprockets and webpacker seems a bad idea: how do I explain to new developers that javascripts are managed by webpacker and stylesheets by sprockets? Seems a bit crazy to me.
I would rather make a cut, take a clear decision, and switch to webpacker completely, rename the `javascripts` folder to `webpacker` (not `assets` to simplify migration process) and have the courage to embrace webpack 100%.

**Comment by dhh** [comment]
If you can't explain to developers that JavaScript is managed by one process and stylesheets by another, I think maybe you're moving too fast in your educational program with these developers. Understanding that JavaScript is separate from stylesheets seems like a pretty fundamental thing to start with before diving into full application development.

But you're also free to change your curriculum to focus on Webpacker. There is support for Webpacker handling other assets in addition to JavaScript, but it's not going to be the default in Rails 6.

**Comment by dhh** [comment]
This branch is now nearly ready to merge, but there are a fair amount of railties tests that need to be updated to match the new generation defaults. If anyone on this thread is interested in helping getting this PR to master, I'd much appreciate some assistance updating these railties tests. You can start a new PR based off the webpacker-default branch.

Once this is merged to master, I'm going to proceed with the reveal of the new Rails 6.0 framework that depends on these changes 😄

**Comment by ahorek** [comment]
```
sprockets are unmaintained at least for a year
sass unmaintained
sassc unmaintained
```

but sprockets is still a hard runtime dependency in rails, I think it should be optional.

My concern is that migration of an existing app from sprockets to full stack webpacker isn't easy, it requires many, many changes. It may even discourage people from upgrading if it would be the only option. We also decided to move to webpacker because working with sprockets was really painful. I support the idea that it should be default even for sass/css.

But step by step, let's finish this PR first, merge it and discuss later.

**Comment by rafaelfranca** [comment]
> sprockets are unmaintained at least for a year

Who said that? We just released a new 4.0 beta 28 days ago.

**Comment by ahorek** [comment]
well, sprockets 4 is in beta since February 18, 2016 and I don't believe it'll ever be released in stable. Thanks for your work on it, but AFAIK you're the only maintainer.

**Comment by dhh** [comment]
This is unproductive. Please refrain from any further distraction on this thread that’s unrelated to the scope of this PR: Making Webpacker the default JavaScript compiler for Rails 6.

If you wish to have a conversation beyond that scope, please take it to the mailing list or elsewhere.

**Comment by guilleiguaran** [comment]
Update: after of merging the PRs of @ttanimichi  we still having failures in the next files:

```
Failed in:
  test/application/rake_test.rb
  test/generators/controller_generator_test.rb
  test/generators/scaffold_controller_generator_test.rb
  test/generators/app_generator_test.rb
  test/generators/plugin_generator_test.rb
  test/generators/scaffold_generator_test.rb
```

**Comment by dhh** [comment]
Thanks a lot for helping with the test updates, @ttanimichi 🙏

**Comment by ttanimichi** [comment]
My pleasure. I think webpacker is a good compromise to manage modern JS in Rails.

**Comment by kaspth** [comment]
Tried to reduce more of the failures. Several tests spit this out:

```ruby
kaspers-macbook-pro:railties kasperhansen$ bin/test test/generators/app_generator_test.rb -n /test_generator_does_not_generate_active_storage_contents_if_skip_active_record_is_given/
Run options: -n /test_generator_does_not_generate_active_storage_contents_if_skip_active_record_is_given/ --seed 31939

# Running:

rails aborted!
Don't know how to build task 'webpacker:install' (see --tasks)
/Users/kasperhansen/code/rails/railties/lib/rails/commands/rake/rake_command.rb:23:in `block in perform'
/Users/kasperhansen/code/rails/railties/lib/rails/commands/rake/rake_command.rb:20:in `perform'
/Users/kasperhansen/code/rails/railties/lib/rails/command.rb:48:in `invoke'
/Users/kasperhansen/code/rails/railties/lib/rails/commands.rb:18:in `<main>'
/Users/kasperhansen/code/rails/activesupport/lib/active_support/dependencies.rb:287:in `block in require'
/Users/kasperhansen/code/rails/activesupport/lib/active_support/dependencies.rb:253:in `load_dependency'
/Users/kasperhansen/code/rails/activesupport/lib/active_support/dependencies.rb:287:in `require'
bin/rails:4:in `<main>'
(See full trace by running task with --trace)
.

Finished in 2.971897s, 0.3365 runs/s, 7.0662 assertions/s.
1 runs, 21 assertions, 0 failures, 0 errors, 0 skips
```

So it looks like the `TestApp::Application` doesn't have access to `webpacker:install`. I tried adding `webpacker` to the Gemfile with a require in `railties/test/abstract_unit.rb`, but then the test would just hang 😕

`isolation/abstract_unit.rb` likely also needs to require webpacker somehow (in the same vein that it does with Turbolinks).

**Comment by ttanimichi** [comment]
@kaspth Maybe that's because `ENV['BUNDLE_GEMFILE']` is set. We have discussed it in [this thread](https://github.com/rails/rails/pull/32780#discussion_r185307068).

**Comment by lsylvester** [comment]
@dhh Is it possible to get a release of webpacker that includes rails/webpacker@8ca372dc500d16a2343aaade7646647aac6925b3. I think that this is required for the `webpacker:install` commands to run correctly within the generators.

I can currently manually adding it to the test app in https://github.com/rails/rails/pull/33622/files#diff-6b400e02467a1f4e0c2c60863c87b927R471 but this should not be requires and is causing some issues with the generator tests.

**Comment by lsylvester** [comment]
I have found another place where the build can hang. In `bin/rails app:update` it waits for user input  to confirm to overwrite the development.rb and production.rb as they are different because of the code that web packer injects into them. If we include the webpacker config in the templates in the railties then then it can continue without hanging waiting for user input.

File: railties/lib/rails/generators/rails/app/templates/package.json.tt [hunk 2]
```diff
@@ -1,5 +1,10 @@
 {
   "name": "<%= app_name %>",
   "private": true,
-  "dependencies": {}
+  "dependencies": {
+    "rails-ujs": "<%= Rails::VERSION::STRING %>"<% unless options[:skip_turbolinks] %>,
```
Comment (n-rodriguez) [inline railties/lib/rails/generators/rails/app/templates/package.json.tt]: Hi there! I think there's something wrong here... `"rails-ujs" unless options[:skip_turbolinks]`, etc...

Comment (georgeclaghorn) [reply railties/lib/rails/generators/rails/app/templates/package.json.tt]: This is correct: we’re skipping the *next* dependency.

**Comment by lsylvester** [comment]
I am having issues with the build timing out for my branches that are trying to provide fixes for this. 

The new generator step of `webpacker:install` is causing some of the generator tests (I am looking at `bin/test test/generators/plugin_generator_test.rb  -v` currently) to take significantly longer than before (from less than a second to around 20 seconds). From what I can tell, most of this time is running `yarn add` commands and deleting files. 

I am trying to work out ways to speed up the `yarn add` step, but I have had no success so far. If anyone has ideas on how to do this that would be really useful.

Otherwise, would it be acceptable to skip javascript on all the tests for generators  that aren't explicitly trying to do anything with the javascript?

The only other option that I can think of is to split the railties tests on travis into different jobs so that each job can run under 50 minutes.

**Comment by dhh** [comment]
I think it's perfectly fine to turn off javascript generation for tests
that do not explicitly seek to test that.

**Comment by lsylvester** [comment]
#33640 which adds web packer to the gentile without handing/timing out the build is ready for review.

**Comment by lsylvester** [comment]
Don't anyone have any guidance on what the js support for a newly generated plugin should have?

Going forwards, plugins will have to publish the javascript via npm so that they can be used by apps using webpacker. So I think that they will need to have at a minimum a `package.json` and a `app/javascript/%name%/index.js`. 

I think that it would be good if they could be generated with the ability to compile the `app/javascript` to `app/assets/javascripts` so that they could ship the javascript for both webpacker and sprockets applications, but I am not sure what the most appropriate tool to do this would be. Active Storage seems to be using [rollup](https://github.com/rollup/rollup) to do this. Would generating a plugin with a minimal rollup config be a sensible default?

**Comment by dhh** [comment]
Anyone interested in picking up the work on the testing side and push this over the finish line? I have that new framework sitting all pretty waiting for this to land in master 😄

**Comment by lsylvester** [comment]
@dhh Did you have any thoughts on the work that was done in #33640 for this? I think from a build speed point of view (which is there main problem it is facing) I think the most important thing would be to replace the `yarn add` in web packer which direct modification of `package.json` and a skip-able `yarn install`. This would allow for the tests that check the generated files but don't need to actually run the javascript skip interacting with yarn at all, which is the main bottleneck.

**Comment by dhh** [comment]
@lsylvester I hadn't seen that. Merged it in! Do see my note about hardcoding the framework version, though. Fine skipping yarn install for the tests 👍

**Comment by lsylvester** [comment]
#34009 Now has all the tests passing for this branch 🎉 

These is still some more things to clean up - like the removing the hard coded package version numbers and figuring out the js support for engines.

The failure on `test/application/rake_test.rb` that is on this branch I didn't get of the last PR so I think it might be related to cached node_modules or something. The error is `DRb::DRbConnError` which is not super helpful - it is likely it was an error in an template as they can't be marshalled. I am also working on a fix for that.

/cc @dhh

**Comment by dhh** [comment]
@lsylvester Can you clear the codeclimate report issues?

**Comment by lsylvester** [comment]
@dhh looks like @consti beat me to it.

**Comment by dhh** [comment]
There’s still a bunch of failures on the railties suite. Also, for now, we should have an open ended version bind to the npm dependencies. Just the latest released version.

**Comment by prashuchaudhary** [comment]
@TomaszGrudzinski I don't think there is any problem with that commit.  I patched the same thing in a sample app and it works.

 It might be that for some reason definition of `active_support/hash_with_indifferent_access.rb` is conflicting when `rails/webpacker` is requiring `active_support/core_ext/hash/indifferent_access`

**Comment by lsylvester** [comment]
I have a fix for the the `except` issue for `HashWithIndiferentAccess` on #34031. That branch is against master as it is not specific to webpacker. 

Also #34030 should help with the `DRb` errors being reported as it will allow error messages raised from templates be seen with the real error messages.

It would be good to get those merged and master merged back into this branch.

Also #34011 will remove the hard coding of the package versions and is the latest patch release for the package. I am not sure about allowing any of the latest packages as at some point the packages would become incompatible and generating a app in an old version of rails would result in it being in a broken state.

@dhh Are you happy constrained to latest patch version, or do you want it fully opened up. I think at some point we need to have the gems being able to control which version of the package is used - but I haven't had a chance to work on that yet.

**Comment by dhh** [comment]
@lsylvester Awesome. I've merged everything and pushed the latest. Let's see if this makes the CI happy and we can finally merge!

I'm a fan of open dependency versioning unless you know that there's a specific incompatibility you're defending against. So leaving them open.

**Comment by partounian** [comment]
I can't wait to upgrade from Rails 4 to Rails 6, this ticket is making me hyped

**Comment by n-rodriguez** [comment]
> I can't wait to upgrade from Rails 4 to Rails 6, this ticket is making me hyped

maybe you should migrate to Rails 5 before :)

**Comment by shashiprab** [comment]
@dhh keeping my eyes peeled for the promised new JS framework for Rails 6.0 reveal! (Haven't found anything satisfying so far with either React, or Stimulus or Ember - hope its a new "Rails Way" framework)

**Comment by dhh** [comment]
It's been revealed! https://www.youtube.com/watch?v=HJZ9TnKrt7Q

There's another Rails 6 reveal in the wings still, but that's not a JS
related framework.

**Pull request merged**